### PR TITLE
CBM installation rebalance take 3

### PR DESCRIPTION
**Pull request by Firestorm01X2**
CBM installation rebalance take 3

#### Summary
SUMMARY: Balance "CBM installation rebalance"

#### Purpose of change
Making  early CBM installation viable option, but with possible consequences.
Working towards:
https://github.com/cataclysmbnteam/Cataclysm-BN/projects/3

#### Describe the solution
unknown

1) CBM instllation is now guranteed, but after that failure effects still may appear. So even failed operation gives bionic to player in workable condition. No more wasted bionic, but you still need skills if you want to install bionics without injures or failed bionics in addition.
2) CBM uninstallation works as usual- it can fail completely.
3) Autodoc now can't kill player during operation due damaging body parts. Instead of finally killing player it just add infection on body or head. It still can break limbs, however.
4) Autodoc damage to player body parts now heavily depends on bionic difficulty. So failures on low level bionics during installation or uninstallation should not critically damage player.
5) Autodoc specialists and MD's now has more benefits since they can't install failed bionics - even in worse case scenarious they just get damage, but because of chages they always gets bionics in workable.
6) Autodoc faulires no longer causes pain. Since palyer is on narcosis during operation anyway it was useless.

* [x] Making installation guranteed but with failure chances after installation.
* [x] Update UI messages to reflect mechanics change.
* [x]  Remove pain effect on cbm failures- player is on narcosis anyway and narcosis blocks pain for player already.
* [x] Reduce body part damage during operation fails to reduce chance for player to outright kill himself.

#### Testing
1) Try to install CBM. 
2) Installation should always happened.
3) Players should not be able to die during operation.
4) There is no pain case should appers.
5) Uninstallation should work as usual.

#### Additional context
Saves for testing:
[CBM.zip](https://github.com/cataclysmbnteam/Cataclysm-BN/files/5845010/CBM.zip)

### CODE REVIEW

File: src/bionics.cpp [hunk 1]
```diff
@@ -2236,29 +2236,29 @@ void Character::perform_install( bionic_id bid, bionic_id upbid, int difficulty,
...
-        add_bionic( bid );
+    g->events().send<event_type::installs_cbm>( getID(), bid );
+    if( upbid != bionic_id( "" ) ) {
+        remove_bionic( upbid );
+        //~ %1$s - name of the bionic to be upgraded (inferior), %2$s - name of the upgraded bionic (superior).
+        add_msg( m_good, _( "Successfully upgraded %1$s to %2$s." ),
+                 upbid.obj().name, bid.obj().name );
+    } else {
+        //~ %s - name of the bionic.
+        add_msg( m_good, _( "Successfully installed %s." ), bid.obj().name );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2248]: If all installations are successful, there should be no "successfully". Just "installed %s", followed by "installation completed with no issues" or whatever issue happened.

File: src/bionics.cpp [hunk 2]
```diff
@@ -2284,7 +2284,7 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
     int failure_level = static_cast<int>( std::sqrt( success * 4.0 * difficulty / adjusted_skill ) );
     int fail_type = ( failure_level > 5 ? 5 : failure_level );
     bool drop_cbm = false;
-    add_msg( m_neutral, _( "The installation is a failure." ) );
+    add_msg( m_bad, _( "Partial failure happened during installation!" ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2287]: Should no longer be called a failure, but an issue. Unless it actually fails to install the bionic.

File: src/bionics.cpp [hunk 3]
```diff
@@ -2356,18 +2357,22 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
                     const bionic_id &id = random_entry( valid );
                     add_bionic( id );
                     g->events().send<event_type::installs_faulty_cbm>( getID(), id );
+                    add_msg( m_bad, _( "Faulty CBM part installed - %s.  Unistall it to clear the malfunction." ),
```
Comment (Coolthulhu) [inline src/bionics.cpp:2360]: "CBM" is the item which contains a bionic, bionic itself is no longer a CBM.

It shouldn't imply that the faulty bionic is a separate part. After this PR, the faulty bionic will be an abstraction, a representation of a flaw in the installed bionic, caused by incorrect installation or damage during installation.
I'm not sure how to reword it properly, though.

**Comment by Coolthulhu** [comment]
> Bonus to skill from MD professions increased approximately at 3 times (bonus to pl_skill from +3 to +10) to make it more impactful.

It prevents harder complications from installation, though. It's already a noticeable bonus.

**Comment by Firestorm01X2** [comment]
> It prevents harder complications from installation, though. It's already a noticeable bonus.

Agreed. Considering changes further buffing is unnnecesary.

File: src/bionics.cpp (hunk 3 above)
Comment (Firestorm01X2) [reply src/bionics.cpp]: Probably "bionic malfunction" would be good synonym

**Comment by Firestorm01X2** [comment]
Ok. Few new things.
1) Pain during operation faulire actually completely ignored for player.
mod_pain() for player calls player::modpain() that actually ignores pain if player has narcosis status. It is totally make sense but that makes pain fail effect useless
2) Severe limb damage due operation fail way too severe and can be up to 80 hp. It is outright could kill player. I think it should be reduced.

**Comment by Firestorm01X2** [comment]
> That's stupid, narcosis shouldn't prevent feeling pain after the operation. If you remove that part form `mod_pain`, does anything weird happen?

```
void player::mod_pain( int npain )
{
    if( npain > 0 ) {
        if( has_trait( trait_NOPAIN ) || **has_effect( effect_narcosis )** ) {
            return;
        }
        // always increase pain gained by one from these bad mutations
        if( has_trait( trait_MOREPAIN ) ) {
            npain += std::max( 1, roll_remainder( npain * 0.25 ) );
        } else if( has_trait( trait_MOREPAIN2 ) ) {
            npain += std::max( 1, roll_remainder( npain * 0.5 ) );
        } else if( has_trait( trait_MOREPAIN3 ) ) {
            npain += std::max( 1, roll_remainder( npain * 1.0 ) );
        }

        if( npain > 1 ) {
            // if it's 1 it'll just become 0, which is bad
            if( has_trait( trait_PAINRESIST_TROGLO ) ) {
                npain = roll_remainder( npain * 0.5 );
            } else if( has_trait( trait_PAINRESIST ) ) {
                npain = roll_remainder( npain * 0.67 );
            }
        }
    }
    Creature::mod_pain( npain );
}
```

**Comment by Firestorm01X2** [comment]
>     * If cap was reached for head or torso, add an infection on one of those parts (just one)

That can be quite harsh also and may lead to conlsusion for player- do not attempt to operate autodoc without antibiotics.
Also considering current logic body part damage is should be less worse than failed bionic installation. But that way less worse outcome has potential to kill players but more worse- just adds long term effect.

**Comment by Coolthulhu** [comment]
> That can be quite harsh also and may lead to conlsusion for player- do not attempt to operate autodoc without antibiotics.

It only covers a small subset of characters - ones with low hp - and isn't all that harsh. It also works better for frail characters, who, in your version, can still die.
> just adds long term effect

It's harder to remove a faulty bionic than an infection. Especially at the point you're already doing CBM installs, when you should have or be able to quickly acquire a stack of antibiotics.

**Comment by Coolthulhu** [comment]
> It simply prevents setting it.

Yeah, I mean, does anything actually weird, like aborting the surgery, happen with it removed?

**Comment by Firestorm01X2** [comment]
> Yeah, I mean, does anything actually weird, like aborting the surgery, happen with it removed?

I've tried to set pain using `void Creature::mod_pain_noresist( int npain )` and it's worked. But it asked player to stop operation after getting pain. Anyway even after answered yes player still continue sleeping.
Probably it could be avoided if I block `on_stat_change( "pain", pain );` via parameter.

**Comment by Coolthulhu** [comment]
> Probably it could be avoided if I block on_stat_change( "pain", pain ); via parameter.

It doesn't look like the right place.
I don't think the pain effect is worth keeping, it's a non-issue anyway. We might bring it back when manual installation is properly reintroduced. Until then, it's fine for the narcosis to block all pain.

**Comment by Firestorm01X2** [comment]
> I don't think the pain effect is worth keeping

You mean I should just get rid of pain effect on fails completely or just left it in current state? Obviosly it is just lefovers from naual installation then there was no narcosis.

**Comment by Coolthulhu** [comment]
> You mean I should just get rid of pain effect on fails completely or just left it in current state?

Just remove the "it's painful!" failure and have the narcosis prevent pain. Even if it caused 1000 pain, it would heal so fast that it doesn't matter.

**Comment by Firestorm01X2** [comment]
> Just remove the "it's painful!" failure and have the narcosis prevent pain. Even if it caused 1000 pain, it would heal so fast that it doesn't matter.

Our I just can leave it but add check or narcosis right there. After that it could be used for manual installation.

**Comment by Coolthulhu** [comment]
> Our I just can leave it but add check on narcosis right there.

No, it's trivial to re-add later.
Narcosis needs that anti-pain effect anyway, otherwise the pain effect from damage will interrupt surgery.

File: src/bionics.cpp [hunk 4]
```diff
@@ -2140,7 +2140,7 @@ bool Character::can_install_bionics( const itype &type, player &installer, bool
         }
     } else {
         if( !g->u.query_yn(
-                _( "WARNING: %i percent chance of failure that may result in damage, pain, or a faulty installation!  Continue anyway?" ),
+                _( "WARNING: Installation of bionic is guranteed but there is %i percent chance of issue that may result in damage, pain, or a partially faulty installation!  Continue anyway?" ),
```
Comment (Coolthulhu) [inline src/bionics.cpp:2143]: No need to mention guarantees. Just say
"WARNING: There is a %i percent chance of complications, such as damage or faulty installation"

**Comment by Firestorm01X2** [comment]
> No, it's trivial to re-add later.

There are mods also that coud potentially use manual installation right now.

**Comment by Firestorm01X2** [comment]
> Narcosis needs that anti-pain effect anyway, otherwise the pain effect from damage will interrupt surgery.

Right. Removing

File: src/bionics.cpp [hunk 5]
```diff
@@ -2356,18 +2357,22 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
                     const bionic_id &id = random_entry( valid );
                     add_bionic( id );
                     g->events().send<event_type::installs_faulty_cbm>( getID(), id );
+                    add_msg( m_bad, _( "Bionic malfunction appeared - %s.  Unistall it to clear the malfunction." ),
```
Comment (Coolthulhu) [inline src/bionics.cpp:2344]: "Malfunction" is OK, but "appeared" is weird.
How about "Errors in installation caused a malfunction - %s."

Also, "uninstall" - you missed an n.

File: src/bionics.cpp [hunk 6]
```diff
@@ -2330,7 +2331,7 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
 
             case 4:
             case 5: {
-                add_msg( m_bad, _( "The installation is faulty!" ) );
+                add_msg( m_bad, _( "Malfunction happened during installation!" ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2318]: This is no longer needed. It was useful when the faulty bionic wasn't printed, but now it is.

File: src/bionics.cpp [hunk 7]
```diff
@@ -2297,16 +2297,17 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
         }
     }
     if( fail_type <= 0 ) {
-        add_msg( m_neutral, _( "The installation fails without incident." ) );
+        add_msg( m_neutral, _( "The installation issue ended up without serious incidents." ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2293]: This needs to be merged with the case when `success > 0`. Either both of those should print the same message or neither should print a message.

File: src/bionics.cpp [hunk 8]
```diff
@@ -2284,7 +2284,7 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
     int failure_level = static_cast<int>( std::sqrt( success * 4.0 * difficulty / adjusted_skill ) );
     int fail_type = ( failure_level > 5 ? 5 : failure_level );
     bool drop_cbm = false;
-    add_msg( m_neutral, _( "The installation is a failure." ) );
+    add_msg( m_bad, _( "Issue happened during installation!" ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2280]: Can probably be removed

File: src/bionics.cpp (hunk 7 above)
Comment (Firestorm01X2) [reply src/bionics.cpp:2293]: Already done.
` if( fail_type <= 1 ) {
        add_msg( m_neutral, _( "The installation issue ended up without serious incidents." ) );
        drop_cbm = true;
    } else {`

Comment (Coolthulhu) [reply src/bionics.cpp:2293]: Nope: you print it only when `success <= 0` in the calling function.
The two cases - "fully successful installation" and "successful installation that looks like it could be bad, but isn't" - are one and the same and need to have same message. Or lack of message.

**Comment by Firestorm01X2** [comment]
Probably it will be good idea to attach part damage to bionic difficulty. Like max_damage=diffuculty*14.

File: src/iexamine.cpp [hunk 9]
```diff
@@ -4518,7 +4518,7 @@ void iexamine::autodoc( player &p, const tripoint &examp )
         const std::string &warning = warning_sign + colorize( _( " WARNING: Operator missing" ),
                                      c_red ) + warning_sign;
         autodoc_header = warning +
-                         _( "\n Using the Autodoc without an operator can lead to <color_light_cyan>serious injuries</color> or <color_light_cyan>death</color>.\n By continuing with the operation you accept the risks and acknowledge that you will not take any legal actions against this facility in case of an accident. " );
+                         _( "\n Using the Autodoc without an operator can lead to <color_light_cyan>serious injuries</color> or <color_light_cyan>death</color>. \n Manufacturer <color_light_green>guarantees CBM installation in workable condition</color> even for incorrect operation programming, but sewere <color_light_cyan>issues</color> may appear. \n Manufacturer <color_light_cyan>does not guarantee sucessfull CBM uninstallation</color> due incorrect opration programming. \n By continuing with the operation you accept the risks and acknowledge that you will not take any legal actions against this facility in case of an accident. " );
```
Comment (Coolthulhu) [inline src/iexamine.cpp:4521]: * "Severe" - 'w'->'v'
* Sucessfull -> Successful
* I'm not sure about this addition. The "guaranteed to work when installed" part alone could be good, but the warning afterwards doesn't sound "professional". I'm not sure if this part is needed (UI will show it anyway) and I don't know how to rephrase it to sound right.

File: src/game_inventory.cpp [hunk 10]
```diff
@@ -1616,7 +1616,7 @@ class bionic_install_preset: public inventory_selector_preset
             p( pl ), pa( patient ) {
             append_cell( [ this ]( const item_location & loc ) {
                 return get_failure_chance( loc );
-            }, _( "FAILURE CHANCE" ) );
+            }, _( "ISSUE CHANCE" ) );
```
Comment (Coolthulhu) [inline src/game_inventory.cpp:1619]: Complication chance? Malfunction chance?
"Issue" is rather ambiguous in this case. I guess it's also ambiguous in other cases.

File: src/bionics.cpp [hunk 11]
```diff
@@ -2270,6 +2206,42 @@ void Character::perform_install( bionic_id bid, bionic_id upbid, int difficulty,
...
 }
 
+void Character::do_damage_for_bionic_failure( int min_damage, int max_damage )
+{
+    std::set<body_part> bp_hurt;
+    for( const bodypart_id &bp : get_all_body_parts() ) {
+        const body_part enum_bp = bp->token;
+        const hp_part hppart = bp_to_hp( enum_bp );
+        bool infection_added = false;
+        if( has_effect( effect_under_op, enum_bp ) && enum_bp != num_bp ) {
```
Comment (Coolthulhu) [inline src/bionics.cpp:2216]: That `num_bp` rearing its ugly head again. I'll have to remove it someday, from what I've seen, DDA did it.
Still, it's not your fault and it probably needs to be here.

File: src/bionics.cpp [hunk 12]
```diff
@@ -2270,6 +2206,42 @@ void Character::perform_install( bionic_id bid, bionic_id upbid, int difficulty,
...
+            bp_hurt.emplace( mutate_to_main_part( enum_bp ) );
+
+            int damage = rng( min_damage, max_damage );
+            int hp = get_hp( hppart );
+            if( hp > damage || ( hppart != bp_head && hppart != bp_torso ) ) {
+                apply_damage( this, bp, damage, true );
+                if( damage > 15 )
+                    add_msg_player_or_npc( m_bad, _( "Your %s is severly damaged." ),
+                                           _( "<npcname>'s %s is severly damaged." ),
+                                           body_part_name_accusative( enum_bp ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2229]: "Severely"

File: src/bionics.cpp [hunk 13]
```diff
@@ -2270,6 +2206,42 @@ void Character::perform_install( bionic_id bid, bionic_id upbid, int difficulty,
...
+        bool infection_added = false;
+        if( has_effect( effect_under_op, enum_bp ) && enum_bp != num_bp ) {
+            if( bp_hurt.count( mutate_to_main_part( enum_bp ) ) > 0 ) {
+                continue;
+            }
+            bp_hurt.emplace( mutate_to_main_part( enum_bp ) );
+
+            int damage = rng( min_damage, max_damage );
+            int hp = get_hp( hppart );
+            if( hp > damage || ( hppart != bp_head && hppart != bp_torso ) ) {
```
Comment (Coolthulhu) [inline src/bionics.cpp:2224]: Would be good if head and torso got damaged too, just less than 100%.
Also, enum mismatch warning: you're comparing `hppart`, which is a (duh) hp_part, with `bp_head/torso`, which are `bp_part`. They are compared as `int`s so it may compile, but it won't work. You need to compare with `hp_head/torso`.

I'd do it like this:
```
if( damage >= hp && ( hppart == bp_head || hp_part == bp_torso ) ) {
  infection_added = true;
  // Rest of the infection stuff
  damage *= 0.8f;
}

apply_damage( // Rest of the stuff, this time unconditionally
```

**Comment by Firestorm01X2** [comment]
It seems that with hight failure chance you always got faulty bionics, without chance to stay with just minor damage. Tried multiple times with 99% failure chance it always got me faulty bionics. It can be fixed with extra random.

**Comment by Coolthulhu** [comment]
> It can be fixed with extra random.

I don't think it's needed. Faulty bionics are worse than minor damage and so are better for balancing out risky installs.

**Comment by Firestorm01X2** [comment]
> I don't think it's needed. Faulty bionics are worse than minor damage and so are better for balancing out risky installs.

Problem that at 99% failure chance you always get faulty bionis. Always exactly this case happens. Tested a lot of times.

**Comment by Coolthulhu** [comment]
> Problem that at 99% failure chance you always get faulty bionis.

But it's the worst one and it should happen.
If you think that's bad, you could allow more than one problem per install, with the extra problems being of the lower grades.

**Comment by Firestorm01X2** [comment]
> But it's the worst one and it should happen.

Well at 60% it does become better and I've finally catched other cases. Still this way you can't get away with really risky install without consequences.
So, probably most players still going to prefer grind skills and delay intallation.

**Comment by Coolthulhu** [comment]
> So, probably most players still going to prefer grind skills and delay intallation.

Or they can install now and reap the big benefits like +2 stat, then remove the fault once they grind the skill.
Most faulty bionics aren't hard to deal with.

**Comment by Firestorm01X2** [comment]
> Or they can install now and reap the big benefits like +2 stat, then remove the fault once they grind the skill.

It is also make sense. Let'e see how it turns out.

File: src/bionics.cpp (hunk 4 above)
Comment (Firestorm01X2) [reply src/bionics.cpp:2079]: Meaning guarantees made to explain for player that installation of CBM is always happens. If we simplify it- probably players is not eve going to understand that now it is the case.

File: src/bionics.cpp (hunk 11 above)
Comment (Firestorm01X2) [reply src/bionics.cpp:2215]: Also I am not sure why even 
```
std::set<body_part> bp_hurt;
...
bp_hurt.emplace( mutate_to_main_part( enum_bp ) );
```
needed there. But it was there before me. So I've kept it.

File: src/bionics.cpp (hunk 4 above)
Comment (Coolthulhu) [reply src/bionics.cpp:2079]: No need to inform the player at every step. The player will need to learn what happens on their own anyway.
"Faulty" implies "partial". A "fully faulty" would be "broken".

File: src/bionics.cpp (hunk 11 above)
Comment (Coolthulhu) [reply src/bionics.cpp:2215]: It was to prevent double damage to body part with one hp part but more bp parts. For example, the hp part for both hand and arm is arm.

File: src/game_inventory.cpp (hunk 10 above)
Comment (Firestorm01X2) [reply src/game_inventory.cpp:1619]: Errors?

Comment (Coolthulhu) [reply src/game_inventory.cpp:1619]: Errors sound more like a code thing, like "exceptions".
If "malfunction" is "fault", then "complication" sounds like the best option here.

File: src/bionics.cpp (hunk 4 above)
Comment (Firestorm01X2) [reply src/bionics.cpp]: Rephrased

File: src/bionics.cpp (hunk 8 above)
Comment (Firestorm01X2) [reply src/bionics.cpp]: I recommend to keep as in unisntallation as it is showing the fact that something wrong happened.

Comment (Firestorm01X2) [reply src/bionics.cpp]: Rephrased as error

File: src/iexamine.cpp (hunk 9 above)
Comment (Firestorm01X2) [reply src/iexamine.cpp]: Rephrased to be more "serious".

File: src/game_inventory.cpp (hunk 10 above)
Comment (Firestorm01X2) [reply src/game_inventory.cpp]: > Errors sound more like a code thing, like "exceptions".

That is surpsingly make sense since player is programming the thing. But complication sounds more explanable for usual player. Let's do it that way.

File: src/bionics.cpp (hunk 8 above)
Comment (Firestorm01X2) [reply src/bionics.cpp]: Rephrased as error as complication. Or... get rid after all?

File: src/bionics.cpp [hunk 14]
```diff
@@ -2270,6 +2205,41 @@ void Character::perform_install( bionic_id bid, bionic_id upbid, int difficulty,
...
+                damage = hp * 0.8f;
+            }
+            apply_damage( this, bp, damage, true );
+            if( damage > 15 )
+                add_msg_player_or_npc( m_bad, _( "Your %s is severely damaged." ),
+                                       _( "<npcname>'s %s is severely damaged." ),
+                                       body_part_name_accusative( enum_bp ) );
+            else
+                add_msg_player_or_npc( m_bad, _( "Your %s is damaged." ), _( "<npcname>'s %s is damaged." ),
+                                       body_part_name_accusative( enum_bp ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2236]: Make sure to run astyle - missing braces will fail the build otherwise.

File: src/bionics.cpp [hunk 15]
```diff
@@ -2283,8 +2253,7 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
     // are more likely.
     int failure_level = static_cast<int>( std::sqrt( success * 4.0 * difficulty / adjusted_skill ) );
     int fail_type = ( failure_level > 5 ? 5 : failure_level );
-    bool drop_cbm = false;
-    add_msg( m_neutral, _( "The installation is a failure." ) );
+    add_msg( m_bad, _( "Complication happened during installation!" ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2256]: "A complication".
This message is not needed, the complication itself should say that something is wrong.

File: src/bionics.cpp [hunk 16]
```diff
@@ -2296,41 +2265,16 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
             fail_type = rng( 1, 3 );
         }
     }
-    if( fail_type <= 0 ) {
-        add_msg( m_neutral, _( "The installation fails without incident." ) );
-        drop_cbm = true;
+    if( fail_type <= 1 ) {
+        add_msg( m_neutral, _( "The installation complication ended up without serious incidents." ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2269]: This section should be removed.
If nothing happened, then complication didn't happen.

File: src/bionics.cpp (hunk 8 above)
Comment (Coolthulhu) [reply src/bionics.cpp]: I mean, the problems are described by failure events themselves. "You are damaged" etc.
The problem with printing "bad stuff happened" is that there's the medical doctor trait, which makes it possible for bad stuff not to happen. In this case, printing "complications happened" is an error. As in, game error, not "error in installation".

**Comment by Firestorm01X2** [comment]
Ai have to note thar acitivty_handlers stil lcontains mesages about failures that pontentially could be misleading in installation case.

**Comment by Coolthulhu** [comment]
Can you change those?

**Comment by Firestorm01X2** [comment]
> Can you change those?

I guess I can.

File: src/activity_handlers.cpp [hunk 17]
```diff
@@ -3606,22 +3606,41 @@ void activity_handlers::operation_finish( player_activity *act, player *p )
...
     } else {
         if( act->values[1] > 0 ) {
             add_msg( m_good,
                      _( "The operation is a success." ) );
         } else {
-            add_msg( m_bad,
-                     _( "The operation is a failure." ) );
+            if( act->str_values[0] == "install" ) {
+                add_msg( m_bad,
+                         _( "Bionic was installed and activated but complication happened during operation!" ) );
```
Comment (Coolthulhu) [inline src/activity_handlers.cpp:3638]: "a complication"

**Comment by Firestorm01X2** [comment]
Ok. It looks ready. 
Saves attached to simplify for testing.

File: src/bionics.cpp (hunk 14 above)
Comment (Firestorm01X2) [reply src/bionics.cpp:2236]: Asyled

File: src/activity_handlers.cpp [hunk 18]
```diff
@@ -3606,22 +3606,41 @@ void activity_handlers::operation_finish( player_activity *act, player *p )
...
     } else {
         if( act->values[1] > 0 ) {
             add_msg( m_good,
                      _( "The operation is a success." ) );
         } else {
-            add_msg( m_bad,
-                     _( "The operation is a failure." ) );
+            if( act->str_values[0] == "install" ) {
+                add_msg( m_bad,
+                         _( "Bionic was installed and activated but a complication happened during operation!" ) );
```
Comment (Coolthulhu) [inline src/activity_handlers.cpp:3638]: "activated" - does it mean that activated bionics (like power generation or claws) become activated?
If not, it should just mention installation.

`m_warning` would be better than `m_bad`, after all the bionic was actually installed.

**Comment by Firestorm01X2** [comment]
Still, I think chances to install should be  somewhat higher. But it is different case.

File: src/activity_handlers.cpp [hunk 19]
```diff
@@ -3606,22 +3606,41 @@ void activity_handlers::operation_finish( player_activity *act, player *p )
...
-                           _( "a sad beeping noise: \"Operation failed\"" ), true,
-                           "Autodoc",
-                           "failure" );
+            if( act->str_values[0] == "install" ) {
+                add_msg( m_bad,
+                         _( "The Autodoc completes installation and activates bionic but reports about complications during operation." ) );
+                const std::list<tripoint> autodocs = g->m.find_furnitures_with_flag_in_radius( p->pos(), 1,
+                                                     flag_AUTODOC );
+                sounds::sound( autodocs.front(), 10, sounds::sound_t::music,
+                               _( "a sad beeping noise: \"Complications detected! Report to medical personnel immediately!\"" ),
```
Comment (Coolthulhu) [inline src/activity_handlers.cpp:3615]: Double spaces

**Comment by Firestorm01X2** [comment]
Also one extra random at:
void Character::bionics_install_failure( const bionic_id &bid, const std::string &installer,
        int difficulty, int success, float adjusted_skill, const tripoint &patient_pos )
{
    int failure_level = static_cast<int>( std::sqrt( success * 4.0 * difficulty / adjusted_skill ) );
    int fail_type = ( failure_level > 5 ? 5 : failure_level );
    **fail_type = rng( 1, 3 );**
...

Can make things more random even at low chances. Currently low install chances gurantees faulty CBM.

**Comment by Coolthulhu** [comment]
> Still, I think chances to install should be somewhat higher.

After this PR, it will be "chances to install perfectly".
The formula needs rebalancing, but not just a multiplier. If you want higher chances, it would be good if you could present a table of difficulty->chance. Preferably with min and max chances being 0% and 100%, instead of the current soft edges.

I won't be doing a mergetest run today, I'd like to clear that Windows build failure first.

**Comment by Coolthulhu** [comment]
> Can make things more random even at low chances. Currently low install chances gurantees exactly faulty CBM.

It should. Very low chances means long term consequence, every time.
If you want to add multiple failures per operation, it should be a separate PR.

**Comment by Firestorm01X2** [comment]
> I won't be doing a mergetest run today, I'd like to clear that Windows build failure first.

No rush. It can wait as long as it is needed. At least I've finished it.

**Comment by Firestorm01X2** [comment]
I have to note that I have to add one extra message to make messages consistent to end operation messages.

**Comment by Coolthulhu** [comment]
`src/bionics.cpp:2229:14: error: variable ‘infection_added’ set but not used [-Werror=unused-but-set-variable]`

**Comment by Coolthulhu** [comment]
If the damage is enough to kill both head and torso, both will be infected.
The `add_effect` part should be behind an `if`.

**Comment by Firestorm01X2** [comment]
> Alternatively, both could be infected and the check removed. Punishing the player for doing risky surgery like that is fine.

Done this way

File: src/bionics.cpp [hunk 20]
```diff
@@ -2296,78 +2271,52 @@ void Character::bionics_install_failure( const bionic_id &bid, const std::string
...
 
-            case 1:
-                if( !( has_trait( trait_NOPAIN ) ) ) {
-                    add_msg_if_player( m_bad, _( "It really hurts!" ) );
-                    mod_pain( rng( 10, 30 ) );
-                }
-                drop_cbm = true;
-                break;
+    if( fail_type <= 1 ) {
+        add_msg( m_neutral, _( "The complications were minor." ) );
```
Comment (Coolthulhu) [inline src/bionics.cpp:2276]: This has to go.
There were no complications here.
The fastest fix would be to make fail types 0 and 1 do the same as 2 and 3.

Comment (Firestorm01X2) [reply src/bionics.cpp]: Added minor damage for 0 and 1 case.
Ready.

**Comment by Coolthulhu** [comment]
I'm seeing different chances for complications in the:
* CBM selection menu (shows 61% for Power Storage at 8/8/8/8 0 skill)
* WARNING menu (shows 59%)

**Comment by Coolthulhu** [comment]
I see the problem with failed installs always granting failed bionics.
The range for "malfunctioning bionic" is very high compared to other failures.

This is something for other PRs, though.
My ideas:
* Allow multiple failures per operation. Calculate "failure value" like now, then, in loop, apply failure and subtract some of the "failure value" based on failure type (stronger failures count more), until it all runs out. "WARNING: You can get up to 5 failed bionics with this one CBM" would implore one to reconsider.
* Instead of grading failure types, roll one (skipping the ones too strong for weak failure), then make the effect stronger with more failure. High failure damage would cripple your limbs from full hp, high failure mutation would add many bad mutations, high failure failed bionic would add many failed bionics. Weak failures would do what they do now.
* Add harsher failures after failed bionics.
